aips/aip-x-safe-auth-key-mapping.md

```diff
@@ -0,0 +1,143 @@
+---
+aip: (this is determined by the AIP Manager, leave it empty when drafting)
+title: Safe onchain key rotation address mapping for standard accounts
+author: Alex Kahn (43892045+alnoki@users.noreply.github.com)
+discussions-to (*optional): https://github.com/aptos-foundation/AIPs/issues/487
+Status: Draft
+last-call-end-date (*optional): <mm/dd/yyyy the last date to leave feedbacks and reviews>
+type: Standard Framework
+created: 09/17/2024
+updated (*optional): <mm/dd/yyyy>
+requires (*optional): <AIP number(s)>
+---
+
+# AIP-X - Safe onchain key rotation address mapping for standard accounts
+
+## Summary
+
+The onchain key rotation address mapping has functional issues which inhibit
+safe mapping of authentication key to originating address for standard accounts.
+This proposal resolves these issues by adding assorted checks and extra function logic.
+
+### Out of Scope
+
+Note that this AIP does not attempt to address multisig v2 effects, because even
+without the changes in this AIP, it is already possible for a multisig to
+(misleadingly) generate an entry in the `OriginatingAddress` table:
+
+1. Rotate account `A` to have a new authentication key, thus generating an entry
+   in the `OriginatingAddress` table.
+2. Convert account `A` to a multisig via
+   `multisig_account::create_with_existing_account_and_revoke_auth_key`, which
+   will set the account's authentication key to `0x0`, but which will *not*
+   mutate the `OriginatingAddress` table, since it makes an inner call to
+   `account::rotate_authentication_key_internal`.
+3. The `OriginatingAddress` table then (incorrectly) reports that a mapping from
+   the authentication key (from before multisig conversion) to the multisig
+   address.
+
+## Impact
+
+1. Without the changes proposed in this AIP's reference implementation,
+   unproven authentications (specifically those relying on
+   `rotate_authentication_key_call`) will result in an unidentifiable mapping,
+   such that users will be unable to identify accounts secured by their private
+   key unless they have maintained their own offchain mapping. This applies to
+   exotic wallets like passkeys.
+1. The overwrite behavior (described below) for
+   `update_auth_key_and_originating_address_table` can similarly result in an
+   inability to identify an account based on the private key.
+1. A user who authenticates two accounts with the same private key per the below
```

Comment:
> it isn't really supported for multiple keys.

@davidiw I agree that the intended design does not include support for multiple keys, which is exactly the issue: special situations are possible which render the mapping unsafe in practice, hence the guardrails I've added in this AIP

> This actually feels like it is out of scope...

The language in this section is the result of these rounds of feedback from @alinush and @gregnazario:

1. https://github.com/aptos-foundation/AIPs/pull/499#discussion_r1763861781
2. https://github.com/aptos-foundation/AIPs/pull/499#discussion_r1767641518
3. https://github.com/aptos-foundation/AIPs/pull/499#discussion_r1769333977

So seeing as the language specifically addressed these comments I hesitate to remove anything because then I will have invalidated prior rounds of review